src/levanter/models/gemma.py

```diff
@@ -0,0 +1,608 @@
+
+    Gemma's MLP is similar to gemma, with a GeLU activation function.
+    """
+
+    gate_proj: hnn.Linear  # projection from Embed to Mlp
+    up_proj: hnn.Linear  # projection from Embed to Mlp
+    down_proj: hnn.Linear  # projection from Mlp to Embed
+    act: Callable = eqx.static_field()
+
+    @staticmethod
+    def init(
+        Embed: Axis, Mlp: Axis, activation_fn: Union[str, Callable], *, key, use_bias: bool = False
+    ) -> "GemmaMlp":
+        k_fc, k_up_proj, k_down_proj = jrandom.split(key, 3)
+        gate_proj = hnn.Linear.init(Out=Mlp, In=Embed, key=k_fc, use_bias=use_bias)
+        up_proj = hnn.Linear.init(Out=Mlp, In=Embed, key=k_up_proj, use_bias=use_bias)
+        down_proj = hnn.Linear.init(Out=Embed, In=Mlp, key=k_down_proj, use_bias=use_bias)
+        if isinstance(activation_fn, str):
+            activation_fn = ACT2FN[activation_fn]
+        act = activation_fn  # type: ignore
+        return GemmaMlp(gate_proj, up_proj, down_proj, act)
+
+    @named_call
+    def __call__(self, x: NamedArray, *, key=None) -> NamedArray:
+        k_gate, k_up, k_down = maybe_rng_split(key, 3)
+        hidden_states = self.gate_proj(x, key=k_gate)
+        hidden_states = self.act(hidden_states)
+        hidden_states = hidden_states * self.up_proj(x, key=k_up)
+        outputs = self.down_proj(hidden_states, key=k_down)
+        return outputs
+
+    def from_state_dict(self, state_dict: StateDict, prefix: Optional[str] = None):
+        # unflatten the linear layers of HF state_dict to match the shape of GemmaMlp
+        d = {}
+        d.update(
+            unflatten_linear_layers(
+                apply_prefix(prefix, "gate_proj"), state_dict, self.gate_proj, out_dims_first_in_dict=True
+            )
+        )
+        d.update(
+            unflatten_linear_layers(
+                apply_prefix(prefix, "up_proj"), state_dict, self.up_proj, out_dims_first_in_dict=True
+            )
+        )
+        d.update(
+            unflatten_linear_layers(
+                apply_prefix(prefix, "down_proj"), state_dict, self.down_proj, out_dims_first_in_dict=True
+            )
+        )
+
+        return super().from_state_dict(d, prefix)
+
+    def update_state_dict(self, state_dict: StateDict, prefix: Optional[str] = None) -> StateDict:
+        my_dict: StateDict = {}
+        super().update_state_dict(my_dict, prefix=prefix)
+
+        my_dict.update(
+            flatten_linear_layers(apply_prefix(prefix, "gate_proj"), self.gate_proj, out_dims_first_in_dict=True)
+        )
+        my_dict.update(
+            flatten_linear_layers(apply_prefix(prefix, "up_proj"), self.up_proj, out_dims_first_in_dict=True)
+        )
+        my_dict.update(
+            flatten_linear_layers(apply_prefix(prefix, "down_proj"), self.down_proj, out_dims_first_in_dict=True)
+        )
+
+        state_dict.update(my_dict)
+        return state_dict
+
+
+class GemmaAttention(StateDictSerializationMixin, eqx.Module):
```

Comment:
this is another one that could just be LlamaAttention? WDYT?

Reply:
Definitely, I'll switch this over to LLama: I started with the copy-paste without realizing how similar the models were.

Reply:
totally makes sense! Yeah at this point there's really just the one transformer architecture with a few random tweaks.